app/state.go

```diff
@@ -1847,3 +1855,207 @@ func setBrokenArtifactFlag(store store.Store, artName string) {
+		}
+
+		log.Errorf("Update control map check failed: %s, retrying...", err.Error())
+		return NewFetchRetryControlMapState(c.wrappedState), false
+	}
+
+	return NewControlMapState(c.wrappedState), false
+}
+
+type fetchRetryControlMapState struct {
+	waitState
+	wrappedState UpdateState
+	retries      int
+}
+
+func NewFetchRetryControlMapState(wrappedState UpdateState) State {
+	return &fetchRetryControlMapState{
+		waitState: waitState{
+			baseState: baseState{
+				id: datastore.MenderStateFetchRetryUpdateControl,
+				t:  ToNone,
+			},
+		},
+		wrappedState: wrappedState,
+	}
+}
+
+func (f *fetchRetryControlMapState) Handle(ctx *StateContext, c Controller) (State, bool) {
+
+	log.Debugf("Handle fetch update control retry state")
+
+	intvl, err := client.GetExponentialBackoffTime(f.retries, c.GetUpdatePollInterval())
+	if err != nil {
+		return NewUpdateErrorState(
+			NewTransientError(err), f.wrappedState.Update()), false
+	}
+
+	f.retries++
+
+	log.Infof("Wait %v before next update control map fetch/update attempt", intvl)
+	return f.Wait(
+		NewFetchControlMapState(f.wrappedState),
+		f,
+		intvl,
+		ctx.WakeupChan,
+	)
+}
+
+type controlMapPauseState struct {
+	waitState
+	wrappedState UpdateState
+}
+
+func NewControlMapPauseState(wrappedState UpdateState) State {
+	return &controlMapPauseState{
+		waitState: waitState{
+			baseState: baseState{
+				id: datastore.MenderStateUpdateControlPause,
+				t:  ToNone,
+			},
+		},
+		wrappedState: wrappedState,
+	}
+}
+
+// Pause until 1 of two events:
+//
+// 1. a timeout, which means we need to refresh the map
+// 2. an update map expires, which means we need to query the next action
+//
+func (c *controlMapPauseState) Handle(ctx *StateContext, controller Controller) (State, bool) {
+
+	log.Debug("Handle controlMapPause state")
+
+	// Schedule a timer for the next update map event to
+	// fetch from the server
+	nextMapRefresh, err := controller.GetControlMapPool().
+		NextControlMapHalfTime(c.wrappedState.Update().ID)
+
+	if errors.Is(err, NoUpdateMapsErr) {
+		log.Error("No control maps no longer present, continuing")
+		nextMapRefresh = time.Now().Add(30 * time.Second)
+	}
+
+	updateMapFromServerIn := nextMapRefresh.Sub(time.Now())
+	log.Infof("Next update refresh from the server in: %s", updateMapFromServerIn)
+
+	if updateMapFromServerIn < 0 {
+		updateMapFromServerIn = 30
+	}
+
+	// wait until further notice
+	log.Debug("Pausing the event loop")
+	c.cancel = controller.GetControlMapPool().Updates
+	noopChan := make(chan bool)
+	// NOTE: Since we are calling cancel here, to succesfully multiplex on
+	// the wait state, we need to ignore the returned boolean value, which,
+	// if returned is true, and will stop the daemon entirely.
+	next, _ := c.Wait(
+		NewFetchControlMapState(c.wrappedState),
+		NewControlMapState(c.wrappedState),
+		updateMapFromServerIn,
+		noopChan, // No-op - using the cancel channel for wake-up
+	)
+	return next, false
```

Comment:
This was my initial idea too, but the `WaitState` is a little odd, in that it returns the same value on both `wakeup`, and timer expiry. So the cancel is the only way to successfully multiplex on the next state to return (see the note above).

Reply:
Hmm, but if you go back to `controlMapState` instead of the `wrappedState`, it will re-check all the maps, and then you don't need to separate between them, right?

Reply:
Discussed f2f, creating a separate wait state for the update control maps, which does not need the cancel hack